safe_url_check/lib/safe_url_check.dart

```diff
@@ -0,0 +1,182 @@
+
+/// Check if [url] is available, without allowing access to private networks.
+///
+/// This will make a `HEAD` request to [url] with given [userAgent], following
+/// redirects if the hostname of URLs do not resolve to a private IP address.
+///
+/// Each I/O is set to timeout after [timeout] duration. This coupled with the
+/// large number of retries allow means that this operation can be slow.
+///
+/// It is good practice to set a custom [userAgent], this allows servers to see
+/// which bots are requesting their resources. This package defaults to the
+/// following user-agent:
+/// ```
+/// User-Agent: package:safe_url_check/1.0.0 (+https://github.com/google/dart-neats/tree/master/safe_url_check)
+/// ```
+Future<bool> safeUrlCheck(
+  Uri url, {
+  int maxRedirects = 8,
+  String userAgent = _defaultUserAgent,
+  HttpClient client,
+  RetryOptions retryOptions = const RetryOptions(),
+  Duration timeout = const Duration(seconds: 90),
+}) async {
+  ArgumentError.checkNotNull(url, 'url');
+  ArgumentError.checkNotNull(maxRedirects, 'maxRedirects');
+  ArgumentError.checkNotNull(userAgent, 'userAgent');
+  ArgumentError.checkNotNull(retryOptions, 'retryOptions');
+  if (maxRedirects < 0) {
+    throw ArgumentError.value(
+      maxRedirects,
+      'maxRedirects',
+      'must be a positive integer',
+    );
+  }
+
+  try {
+    // Always call with a client
+    if (client == null) {
+      final c = HttpClient();
+      try {
+        return await _safeUrlCheck(
+          url,
+          maxRedirects,
+          c,
+          userAgent,
+          retryOptions,
+          timeout,
+        );
+      } finally {
+        c.close(force: true);
+      }
+    }
+
+    return await _safeUrlCheck(
+      url,
+      maxRedirects,
+      client,
+      userAgent,
+      retryOptions,
+      timeout,
+    );
+  } on Exception {
+    return false;
+  }
+}
+
+Future<bool> _safeUrlCheck(
+  Uri url,
+  int maxRedirects,
+  HttpClient client,
+  String userAgent,
+  RetryOptions retryOptions,
+  Duration timeout,
+) async {
+  assert(maxRedirects >= 0);
+  assert(client != null);
+  assert(url != null);
+
+  // If no scheme or not http or https, we fail.
+  if (!url.hasScheme || (!url.isScheme('http') && !url.isScheme('https'))) {
+    return false;
+  }
+
+  final ips = await retryOptions.retry(() async {
+    final ips = await InternetAddress.lookup(url.host).timeout(timeout);
+    if (ips == null || ips.isEmpty) {
+      throw Exception('DNS resolution failed');
+    }
+    return ips;
+  });
+  for (final ip in ips) {
+    // If given a loopback, linklocal or multicast IP, return false
+    if (ip.isLoopback ||
+        ip.isLinkLocal ||
+        ip.isMulticast ||
+        isPrivateIpV4(ip) ||
+        isUniqueLocalIpV6(ip)) {
+      return false;
+    }
+  }
+
+  final response = await retryOptions.retry(() async {
+    // We can't use the HttpClient from dart:io with a custom socket, so instead
```

Comment:
What would that look like? Is there an open bug for that?

Reply:
I didn't check. I checked with `dart:io` owner that there is no way to do this.

We have it for `http2`, but I'm not sure how it would look in `dart:io`, and much less certain it could even be added without breaking compatibility.